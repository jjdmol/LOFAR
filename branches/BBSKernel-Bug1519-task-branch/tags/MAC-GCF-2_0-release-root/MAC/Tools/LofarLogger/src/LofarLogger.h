//#  LofarLogger.h: logger wrapper
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_LOGGER_H
#define LOFAR_LOGGER_H

#include <Common/lofar_string.h>
#include <Common/lofar_map.h>
#include <log4cplus/logger.h>
#include "LofarLoggerNames.h"

// forward declaration

// define convenience macros
#define LOFAR_LOG_TRACE_SCOPE(logger,logEvent) \
    LofarScopeTraceLogger _scopeLogger(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_TRACE(logger,logEvent) \
    LofarLogger::getInstance().trace(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_DEBUG(logger,logEvent) \
    LofarLogger::getInstance().debug(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_INFO(logger,logEvent)  \
    LofarLogger::getInstance().info(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_WARN(logger,logEvent)  \
    LofarLogger::getInstance().warn(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_ERROR(logger,logEvent) \
    LofarLogger::getInstance().error(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_FATAL(logger,logEvent) \
    LofarLogger::getInstance().fatal(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);

/**
 *
 * The LofarLogger class wraps the log4cplus::Logger class
 *
 * Usage:
 *
 *    #define EXAMPLE_LOGGER_NAME "ASTRON.LOFAR.MAC.ExampleLoggerName"
 *
 *    LOFAR_LOG_TRACE(EXAMPLE_LOGGER_NAME,("example %d level message %d","trace",15))
 *    LOFAR_LOG_FATAL(EXAMPLE_LOGGER_NAME,("example fatal level message"))
 *
 * NOTE: the second argument of the macro HAS to be surrounded with brackets.
 */
class LofarLogger
{
    public:

        /**
        * destructor
        */
        virtual ~LofarLogger();

        static LofarLogger getInstance();

        /**
        * logging methods4
        */
        void trace(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void debug(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void info(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void warn(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void error(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void fatal(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);

        log4cplus::Logger& getLogger(const string& logger);
        static string formatString(char* format,...);

    protected:
        /**4
        * constructor. Protected because public construction is not allowed
        */
        LofarLogger();
        LofarLogger(const LofarLogger&);

    private:

        /**
        * Don't allow copying of the LofarLogger object.
        */
        LofarLogger& operator=(const LofarLogger&);

        typedef std::map<const std::string,log4cplus::Logger> TLoggerMap;

        /**
        * keep the loggers in a map for re-use
        */
        TLoggerMap         _loggerMap;
        static LofarLogger _sLofarLogger;
};

/**
 *
 * The LofarScopeTraceLogger class logs trace messages in its constructor and destructor
 *
 */
class LofarScopeTraceLogger
{
    public:
       explicit LofarScopeTraceLogger(const std::string& logger, const std::string& logEvent,const char* file=NULL,int line=-1) :
                    _traceLogger(LofarLogger::getInstance().getLogger(logger),logEvent,file,line)
                {
                };
                ~LofarScopeTraceLogger()
                {
                };

    protected:
                /**
                * constructor. Protected because public construction is not allowed
                */
                LofarScopeTraceLogger();
                LofarScopeTraceLogger(const LofarScopeTraceLogger&);
                LofarScopeTraceLogger& operator=(const LofarScopeTraceLogger&);

    private:
                log4cplus::TraceLogger _traceLogger;
};

#endif
