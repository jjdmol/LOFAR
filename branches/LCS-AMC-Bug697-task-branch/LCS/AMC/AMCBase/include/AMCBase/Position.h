//# Position.h: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_AMCBASE_EARTHCOORD_H
#define LOFAR_AMCBASE_EARTHCOORD_H

// \file
// Class to hold an earth coordinate as lon,lat,height

//# Includes
#include <Common/lofar_iosfwd.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>

namespace LOFAR
{
  namespace AMC
  {

    // \addtogroup AMCBase
    // @{

    // This class represents a position on earth. The position is stored using
    // cartesian coordinates. A position can be constructed from longitude,
    // latitude and height, or from a vector containing cartesian coordinates
    // (x,y,z). The context where the object is used defines the coordinate
    // system and frame, so the class can be used in any kind of frame (like
    // ITRF and geocentric). The correct interpretation of the coordinates
    // should be done by the user of this class.
    class Position
    {
    public:
      // Types of position. 
      enum Types {
        INVALID = -1,   ///< Used when specified value is out of range.
        ITRF,
        WGS84,
        //# Insert new types HERE !!
        N_Types         ///< Number of reference types.
      };

      // Default constructor uses 0 for longitude, latitude and height, and
      // ITRF as reference type.
      Position();

      // The destructor must be virtual destructor, because Direction is
      // derived from Position.
      virtual ~Position() {}

      // Create a position by giving the longitude \a lon and latitude \a lat
      // in radians and the \a h in meters. Reference type can be either ITRF
      // (default), or WGS84.
      Position(double lon, double lat, double h, Types typ = ITRF);

      // Create a position from the cartesian coordinates \a xyz and the
      // reference type \a typ.
      explicit Position(const vector<double>& xyz, Types typ = ITRF);
      
      // Return the position in cartesian coordinates.
      const vector<double>& get() const
      { return itsXYZ; }

      // Return the longitude in radians.
      double longitude() const;
      
      // Return the latitude in radians.
      double latitude() const;
      
      // Return the height in meters.
      double height() const;
      
      // Calculate the dot product of \c this and \a that.
      // \note We could have defined a global operator*() taking two Positions
      // as argument, however, then we wouldn't be able to calculate the dot
      // product of a Postion and a Direction.
      double operator*(const Position& that) const;

      // Return whether position type is valid.
      virtual bool isValid() const
      { return itsType != INVALID; }

      // Return the position type as an \c int.
      virtual int type() const
      { return itsType; }

      // Return the position type as a string.
      virtual const string& showType() const;

    protected:
      // Normalize the position vector to unit length. A Position need not be
      // normalized, but a Direction does. Therefore, this method was made
      // protected.
      // \return The original length of the position vector.
      double normalize();

    private:
      // Position is stored internally in cartesian coordinates.
      vector<double> itsXYZ;

      // Type of earth coordinate.
      Types itsType;
    };

    // Output a psition in ASCII format.
    ostream& operator<< (ostream&, const Position&);

    // Compare two positions for equality.
    // \note Two invalid positions can \e never be equal.
    bool operator==(const Position& lhs, const Position& rhs);

    // @}

  } // namespace AMC

} // namespace LOFAR

#endif
