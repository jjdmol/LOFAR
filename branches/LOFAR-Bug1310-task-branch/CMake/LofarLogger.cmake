#  LofarLogger.cmake: Check for logger (log4cplus, log4cxx or none)
#
#  Copyright (C) 2008-2009
#  ASTRON (Netherlands Foundation for Research in Astronomy)
#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#  $Id$

include(LofarFindPackage)

macro(lofar_logger)

  if(USE_LOG4CPLUS AND USE_LOG4CXX)
    message(FATAL_ERROR "Cannot use Log4Cplus and Log4Cxx at the same time")
  endif(USE_LOG4CPLUS AND USE_LOG4CXX)

  if(USE_LOG4CPLUS)
    MESSAGE(STATUS "Checking for Log4Cplus ...")
    lofar_find_package(Log4Cplus 1 log4cplus/logger.h "log4cplus jpeg")
  endif(USE_LOG4CPLUS)

  if(USE_LOG4CXX)
    MESSAGE(STATUS "Checking for Log4Cxx ...")
    lofar_find_package(Log4Cxx 1 log4cxx/logger.h)
  endif(USE_LOG4CXX)

endmacro(lofar_logger)
