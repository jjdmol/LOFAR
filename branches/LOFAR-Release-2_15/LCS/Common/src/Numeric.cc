//# Numeric.cc: 
//#
//# Copyright (C) 2002-2004
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <Common/Numeric.h>
#include <cstdlib>

namespace LOFAR
{
  bool Numeric::isFinite(float f)
  {
    floatUnion_t mask = { f };
    return (mask.mask & floatExponentMask) != floatExponentMask;
  }

  bool Numeric::isFinite(double d)
  {
    doubleUnion_t mask = { d };
    return (mask.mask & doubleExponentMask) != doubleExponentMask;
  }

  bool Numeric::isNegative(float f)
  {
    floatUnion_t mask = { f };
    return (mask.mask & floatNegativeMask) == floatNegativeMask;
  }

  bool Numeric::isNegative(double d)
  {
    doubleUnion_t mask = { d };
    return (mask.mask & doubleNegativeMask) == doubleNegativeMask;
  }

  bool Numeric::isInf(float f)
  {
    floatUnion_t mask = { f };
    return !isFinite(f) && (mask.mask & floatMantissaMask) == 0L;
  }

  bool Numeric::isInf(double d)
  {
    doubleUnion_t mask = { d };
    return !isFinite(d) && (mask.mask & doubleMantissaMask) == 0LL;
  }

  bool Numeric::isNan(float f)
  {
    floatUnion_t mask = { f };
    return !isFinite(f) && (mask.mask & floatMantissaMask) != 0L;
  }

  bool Numeric::isNan(double d)
  {
    doubleUnion_t mask = { d };
    return !isFinite(d) && (mask.mask & doubleMantissaMask) != 0LL;
  }


  bool Numeric::compare(float lhs, float rhs, floatMask_t maxUlps)
  {

#ifdef USE_INFINITYCHECK
    // If either \a lhs or \a rhs is infinite, then return true if they're
    // exactly equal, i.e. they have the same sign.
    if (isInf(lhs) || isInf(rhs)) return lhs == rhs;
#endif

#ifdef USE_NANCHECK
    // Consider \a lhs and \a rhs unequal when either is NaN. 
    if (isNan(lhs) || isNan(rhs)) return false;
#endif

#ifdef USE_SIGNCHECK
    // Consider \a lhs and \a rhs unequal when they have a different
    // sign, even if they both are very close to zero.
    // \note The check for equality is needed because zero and negative zero
    // have different signs but are equal to each other.
    if (isNegative(lhs) != isNegative(rhs)) return lhs == rhs;
#endif

    floatUnion_t mlhs = { lhs };
    floatUnion_t mrhs = { rhs };
    floatMask_t ilhs = mlhs.mask;
    floatMask_t irhs = mrhs.mask;

    // Make \a ilhs and \a irhs lexicographically ordered as twos-complement
    // long.
    if (isNegative(lhs))
      ilhs = ~ilhs + 1;
    else
      ilhs += floatNegativeMask;

    if (isNegative(rhs))
      irhs = ~irhs + 1;
    else
      irhs += floatNegativeMask;

    // If \a ilhs and \a irhs are less than \a maxUlps apart, then \a lhs and
    // \a rhs are considered equal.
    if (ilhs < irhs)
      return irhs - ilhs <= maxUlps;
    else  
      return ilhs - irhs <= maxUlps;
  }


  bool Numeric::compare(double lhs, double rhs, doubleMask_t maxUlps)
  {

#ifdef USE_INFINITYCHECK
    // If either \a lhs or \a rhs is infinite, then return true if they're
    // exactly equal, i.e. they have the same sign.
    if (isInf(lhs) || isInf(rhs)) return lhs == rhs;
#endif

#ifdef USE_NANCHECK
    // Consider \a lhs and \a rhs unequal when either is NaN. 
    if (isNan(lhs) || isNan(rhs)) return false;
#endif

#ifdef USE_SIGNCHECK
    // Consider \a lhs and \a rhs unequal when they have a different
    // sign, even if they both are very close to zero.
    // \note The check for equality is needed because zero and negative zero
    // have different signs but are equal to each other.
    if (isNegative(lhs) != isNegative(rhs)) return lhs == rhs;
#endif

    doubleUnion_t mlhs = { lhs };
    doubleUnion_t mrhs = { rhs };
    doubleMask_t ilhs = mlhs.mask;
    doubleMask_t irhs = mrhs.mask;

    // Make \a ilhs and \a irhs lexicographically ordered
    if (isNegative(lhs))
      ilhs = ~ilhs + 1;
    else
      ilhs += doubleNegativeMask;

    if (isNegative(rhs))
      irhs = ~irhs + 1;
    else
      irhs += doubleNegativeMask;

    // If \a ilhs and \a irhs are less than \a maxUlps apart, then \a lhs and
    // \a rhs are considered equal.
    if (ilhs < irhs)
      return irhs - ilhs <= maxUlps;
    else  
      return ilhs - irhs <= maxUlps;
  }
  

} // namespace LOFAR
