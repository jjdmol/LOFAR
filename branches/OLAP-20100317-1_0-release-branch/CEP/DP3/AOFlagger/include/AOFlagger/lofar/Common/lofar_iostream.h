//# lofar_iostream.h:
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: lofar_iostream.h 14057 2009-09-18 12:26:29Z diepen $

#ifndef LOFAR_COMMON_IOSTREAM_H
#define LOFAR_COMMON_IOSTREAM_H

// \file

#include <Common/lofar_iosfwd.h>
#include <iostream>

namespace LOFAR
{
  using std::istream;
  using std::ostream;
  using std::iostream;
  using std::cin;
  using std::cout;
  using std::cerr;
  using std::clog;
  using std::endl;
  using std::ends;
  using std::flush;
}

#endif
