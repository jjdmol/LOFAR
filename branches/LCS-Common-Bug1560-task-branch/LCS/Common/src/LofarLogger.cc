//# LofarLogger.cc: Interface to the log4cplus logging package
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <cstdio>					// snprintf
#include <unistd.h>					// readlink
#include <libgen.h>				// basename
#include <cstring>
#include <Common/LofarLogger.h>

#if defined(HAVE_LOG4CXX)
# include <log4cxx/ndc.h>
#endif


namespace LOFAR {

//#------------------------- Internal implementation -----------------------------
//
// lofarLoggerInitNode()
//
// Creates a NDC with the text "application@node" and pushes it
// on the NDC stack
void lofarLoggerInitNode(void) {
	const int	MAXLEN = 128;
	int		applNameLen = 0;
	char	hostName [MAXLEN];
	char	applName [MAXLEN];
	char	loggerId [MAXLEN];

	// try to resolve the hostname
	if (gethostname (hostName, MAXLEN-1) < 0) {
		hostName[0]='\0';
	}

	// try to resolve the applicationname
	applNameLen = readlink("/proc/self/exe", applName, MAXLEN-1);
	if (applNameLen >= 0)
		applName[applNameLen] = '\0';
	else
		strcpy (applName, "");
//	}	

	// construct loggerId and register it.
	snprintf(loggerId, MAXLEN-1, "%s@%s", basename(applName), hostName);
#if defined(HAVE_LOG4CPLUS)
	log4cplus::getNDC().push(loggerId);
#elif defined(HAVE_LOG4CXX)
        log4cxx::NDC::push(loggerId);
#endif
}

}	// namespace LOFAR
