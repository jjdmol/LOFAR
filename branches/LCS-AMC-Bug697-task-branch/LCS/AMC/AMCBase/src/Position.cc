//# Position.cc: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/Position.h>
#include <Common/lofar_iostream.h>
#include <Common/lofar_math.h>
#include <Common/LofarLogger.h>
#include <limits>

using namespace std;

namespace LOFAR
{
  namespace AMC
  {

    Position::Position() : 
      itsXYZ(3, 0.0), itsType(ITRF)
    {
    }
    

    Position::Position (double lon, double lat, double h, Types typ) :
      itsXYZ(3, std::numeric_limits<double>::quiet_NaN()), 
      itsType((INVALID < typ && typ < N_Types) ? typ : INVALID) 
    {
      if (isValid()) {
        double tmp = cos(lat);
        itsXYZ[0] = h * cos(lon) * tmp;
        itsXYZ[1] = h * sin(lon) * tmp;
        itsXYZ[2] = h * sin(lat);
      }
    }


    Position::Position(const vector<double>& xyz, Types typ) :
      itsXYZ(3, std::numeric_limits<double>::quiet_NaN()), 
      itsType((INVALID < typ && typ < N_Types) ? typ : INVALID)
    {
      ASSERT(xyz.size() == 3);
      if (isValid()) itsXYZ = xyz;
    }


    double Position::longitude() const
    {
      return atan2(itsXYZ[1], itsXYZ[0]);
    }
    

    double Position::latitude() const
    {
      double h(height());
      if (h == 0) return asin(itsXYZ[2]);
      else return (asin(itsXYZ[2] / h));
    }
    

    double Position::height() const
    {
      return sqrt(itsXYZ[0] * itsXYZ[0] + 
                  itsXYZ[1] * itsXYZ[1] +
                  itsXYZ[2] * itsXYZ[2]);
    }
    

    const string& Position::showType() const
    {
      //# Caution: Always keep this array of strings in sync with the enum
      //#          Types that is defined in the header file!
      static const string types[N_Types+1] = {
        "ITRF",
        "WGS84",
        "<INVALID>"
      };
      if (isValid()) return types[itsType];
      else return types[N_Types];
    }


    ostream& operator<<(ostream& os, const Position& pos)
    {
      if (!pos.isValid()) 
        return os << pos.showType();
      else 
        return os << "["  << pos.longitude() << ", " << pos.latitude() 
                  << ", " << pos.height() << "] (" << pos.showType() << ")";
    }


    bool operator==(const Position& lhs, const Position& rhs)
    {
      return 
        lhs.isValid() && rhs.isValid() &&
        lhs.type()    == rhs.type()    &&
        lhs.get()     == rhs.get();
    }


    double operator*(const Position& lhs, const Position& rhs)
    {
      double result(0);
      vector<double> x(lhs.get());
      vector<double> y(rhs.get());
      ASSERT(x.size() == y.size());
      for (uint i = 0; i < x.size(); ++i) {
        result += x[i] * y[i];
      }
      return result;
    }
    
  } // namespace AMC

} // namespace LOFAR
