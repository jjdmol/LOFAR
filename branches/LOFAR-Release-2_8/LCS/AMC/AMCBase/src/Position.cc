//# Position.cc: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/Position.h>
#include <AMCBase/Direction.h>
#include <AMCBase/Exceptions.h>
#include <Common/lofar_iostream.h>
#include <Common/lofar_math.h>

namespace LOFAR
{
  namespace AMC
  {

    double Earth::flattening()
    {
      return 1.0 / 298.257223563;
    }


    double Earth::equatorialRadius()
    {
      return 6378137.0;
    }


    Coord3D wgs84ToItrf(double lon, double lat, double h)
    {
      double f      = Earth::flattening();        // flattening
      double a      = Earth::equatorialRadius();  // equatorial radius
      double e2     = f*(2-f);                    // square of eccentricity
      double sinlat = sin(lat);                   // sine of latitude
      double coslat = cos(lat);                   // cosine of latitude

      // Radius of curvature in the prime vertical
      double rN = a / sqrt(1 - e2*sinlat*sinlat);

      // Calculate geocentric coordinates
      vector<double> xyz(3);
      xyz[0] = (rN + h) * coslat * cos(lon);
      xyz[1] = (rN + h) * coslat * sin(lon);
      xyz[2] = ((1-e2)*rN + h) * sinlat;

      return Coord3D(xyz);
    }


    //## --------  Public member functions  -------- ##//

    Position::Position() : 
      itsCoord(Coord3D())
    {
    }
    

    Position::Position (double lon, double lat, double h, Types typ)
    {
      switch(typ) {
      case ITRF:
        itsCoord = Coord3D(lon, lat, h);
        break;
      case WGS84:
        itsCoord = wgs84ToItrf(lon, lat, h);
        break;
      default:
        THROW(TypeException, "Invalid type (typ=" << typ << ") specified");
      }
    }


    Position::Position(const Coord3D& coord, Types typ)
    {
      switch(typ) {
      case ITRF:
        itsCoord = coord;
        break;
      case WGS84:
        itsCoord = 
          wgs84ToItrf(coord.longitude(), coord.latitude(), coord.radius());
        break;
      default:
        THROW(TypeException, "Invalid type (typ=" << typ << ") specified");
      }
    }


    Position& Position::operator+=(const Position& that)
    {
      itsCoord += that.itsCoord;
      return *this;
    }


    Position& Position::operator-=(const Position& that)
    {
      itsCoord -= that.itsCoord;
      return *this;
    }


    Position& Position::operator*=(double a)
    {
      itsCoord *= a;
      return *this;
    }


    Position& Position::operator/=(double a)
    {
      itsCoord /= a;
      return *this;
    }


    double Position::operator*(const Position& that) const
    {
      return itsCoord * that.coord();
    }


    double Position::operator*(const Direction& that) const
    {
      if (that.type() != Direction::ITRF) {
        THROW (TypeException, "Direction type must be ITRF");
      }
      return itsCoord * that.coord();
    }


    //## --------  Global functions  -------- ##//

    ostream& operator<<(ostream& os, const Position& pos)
    {
      return os << pos.coord();
    }


    bool operator==(const Position& lhs, const Position& rhs)
    {
      return lhs.coord() == rhs.coord();
    }


  } // namespace AMC

} // namespace LOFAR
