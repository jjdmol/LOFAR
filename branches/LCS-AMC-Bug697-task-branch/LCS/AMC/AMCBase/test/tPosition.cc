//#  tPosition.cc: test program for the Position class
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/Position.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_iostream.h>
#include <Common/Numeric.h>
#include <cmath>

using namespace LOFAR;
using namespace LOFAR::AMC;

int main(int, const char* argv[])
{
  INIT_LOGGER(argv[0]);

  try {

    // A default constructed position
    Position pos0;

    // A "normalized" position in ITRF
    Position pos1(0.25*M_PI, -0.33*M_PI, 1);

    // A "denormalized" position in WGS84; latitude angle is larger than
    // pi/2. As a result, the "normalized" latitude should be pi - 0.75 pi,
    // and the longitude should be -0.67 pi + pi.
    Position pos2(-0.67*M_PI, 0.75*M_PI, 249.98, Position::WGS84);

    // Invalide coordinate type; xyz-coordinates will be NaN.
    Position pos3(0.5*M_PI, 0.2*M_PI, -115.11, 
                   static_cast<Position::Types>(1294));

    // Vector for storing the cartesian coordinates.
    vector<double> p;

    p = pos0.get();
    ASSERT(pos0.isValid() &&
           pos0.longitude() == 0 && 
           pos0.latitude() == 0 && 
           pos0.height() == 0 &&
           pos0.type() == Position::ITRF);
    ASSERT(p[0] == 0 &&
           p[1] == 0 &&
           p[2] == 0);

    p = pos1.get();
    ASSERT(pos1.isValid() &&
           pos1.type() == Position::ITRF &&
           Numeric::compare(pos1.longitude(), 0.25*M_PI) && 
           Numeric::compare(pos1.latitude(), -0.33*M_PI) &&
           Numeric::compare(pos1.height(), 1));
    ASSERT(Numeric::compare(p[0],  0.3599466369818882) &&
           Numeric::compare(p[1],  0.3599466369818882) &&
           Numeric::compare(p[2], -0.8607420270039436));

    p = pos2.get();
    ASSERTSTR(pos2.isValid() &&
           pos2.type() == Position::WGS84 &&
           Numeric::compare(pos2.longitude(), 0.33*M_PI) && 
           Numeric::compare(pos2.latitude(), 0.25*M_PI) &&
           Numeric::compare(pos2.height(), 249.98), pos2);

    ASSERT(Numeric::compare(p[0],  89.9794603127324) &&
           Numeric::compare(p[1], 152.1469583062028) &&
           Numeric::compare(p[2], 176.7625531610132));

    p = pos3.get();
    ASSERT(!pos3.isValid());

    cout << "pos0 = " << pos0 << endl;
    cout << "pos1 = " << pos1 << endl;
    cout << "pos2 = " << pos2 << endl;
    cout << "pos3 = " << pos3 << endl;

  }

  catch (Exception& e) {
    LOG_ERROR_STR(e);
    return 1;
  }

  return 2;  // forces assay to flag missing .stdout file as an error
}
