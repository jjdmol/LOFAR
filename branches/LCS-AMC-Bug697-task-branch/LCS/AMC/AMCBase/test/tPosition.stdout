pos0 = [0, 0, 0] (ITRF)
pos1 = [0.785398, -1.03673, 1] (ITRF)
pos2 = [1.03673, 0.785398, 249.98] (WGS84)
pos3 = <INVALID>
